src/environmentd/src/http/webhook.rs

```diff
@@ -0,0 +1,306 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Helpers for handling events from a Webhook source.
+
+use std::collections::BTreeMap;
+
+use mz_adapter::AdapterError;
+use mz_ore::str::StrExt;
+use mz_repr::adt::jsonb::JsonbPacker;
+use mz_repr::{ColumnType, Datum, Row, ScalarType};
+use mz_storage_client::controller::StorageError;
+
+use anyhow::Context;
+use axum::extract::Path;
+use axum::response::IntoResponse;
+use axum::Extension;
+use bytes::Bytes;
+use http::StatusCode;
+use thiserror::Error;
+
+use crate::http::Delayed;
+
+pub async fn handle_webhook(
+    Extension(client): Extension<Delayed<mz_adapter::Client>>,
+    Path((database, schema, name)): Path<(String, String, String)>,
+    headers: http::HeaderMap,
+    body: Bytes,
+) -> impl IntoResponse {
+    // Get our adapter client.
+    let client = client.await.context("receiving client")?;
+
+    // Get an appender for the provided object, if that object exists.
+    let maybe_appender = client
+        .append_webhook(database.clone(), schema.clone(), name.clone())
+        .await?;
+    let Some((appender, body_ty, header_ty)) = maybe_appender else {
+        let desc = format!("{database}.{schema}.{name}");
+        return Err(WebhookError::NotFound(desc.quoted().to_string()));
+    };
+
+    // Pack our body and headers into a Row.
+    let row = pack_row(body, headers, body_ty, header_ty)?;
+
+    // Send the row to get appended.
+    appender.append(vec![(row, 1)]).await?;
+
+    Ok(())
+}
+
+/// Given the body and headers of a request, pack them into a [`Row`].
+fn pack_row(
+    body: Bytes,
+    headers: http::HeaderMap,
+    body_ty: ColumnType,
+    header_ty: Option<ColumnType>,
+) -> Result<Row, WebhookError> {
+    // If we're including headers then we have two columns.
+    let num_cols = header_ty.as_ref().map(|_| 2).unwrap_or(1);
+
+    // Pack our row.
+    let mut row = Row::with_capacity(num_cols);
+    let mut packer = row.packer();
+
+    // Pack our body into a row.
+    match body_ty.scalar_type {
+        ScalarType::Bytes => packer.push(Datum::Bytes(&body[..])),
+        ty @ ScalarType::String => {
+            let s = std::str::from_utf8(&body).map_err(|m| WebhookError::InvalidBody {
+                ty,
+                msg: m.to_string(),
+            })?;
+            packer.push(Datum::String(s));
+        }
+        ty @ ScalarType::Jsonb => {
+            let jsonb_packer = JsonbPacker::new(&mut packer);
+            jsonb_packer
+                .pack_slice(&body[..])
+                .map_err(|m| WebhookError::InvalidBody {
+                    ty,
+                    msg: m.to_string(),
+                })?;
+        }
+        ty => {
+            Err(anyhow::anyhow!(
+                "Invalid body type for Webhook source: {ty:?}"
+            ))?;
+        }
+    }
+
+    // Pack our headers into a row, if we're supposed to.
+    let mut headers_s = BTreeMap::new();
```

Comment:
Can this whole block be moved into the if so we don't allocate the map if we don't need to?